task/backend/scheduler/time.go

```diff
@@ -0,0 +1,237 @@
+package scheduler
+
+import (
+	"sync"
+	"time"
+)
+
+// Time is an interface to allow us to mock time.
+type Time interface {
+	Now() time.Time
+	Unix(seconds, nanoseconds int64) time.Time
+	NewTimer(d time.Duration) Timer
+	Until(time.Time) time.Duration
+}
+
+type stdTime struct{}
+
+// Now gives us the current time as time.Time would
+func (stdTime) Now() time.Time {
+	return time.Now()
+}
+
+// Unix gives us the time given seconds and nanoseconds.
+func (stdTime) Unix(sec, nsec int64) time.Time {
+	return time.Unix(sec, nsec)
+}
+
+func (stdTime) Until(t time.Time) time.Duration {
+	return time.Until(t)
+}
+
+// NewTimer gives us a Timer that fires after duration d.
+func (stdTime) NewTimer(d time.Duration) Timer {
+	t := time.NewTimer(d)
+	return &stdTimer{*t}
+}
+
+// Timer is an interface to allow us to mock out timers.  It has behavior like time.Timer
+type Timer interface {
+	C() <-chan time.Time
+	Reset(d time.Duration) bool
+	Stop() bool
+}
+
+// stdTimer is a Timer that wraps time.Time.
+type stdTimer struct {
+	time.Timer
+}
+
+// C returns a <-chan time.Time  and can be used much like time.Timer.C.
+func (t *stdTimer) C() <-chan time.Time {
+	return t.Timer.C
+}
+
+// MockTime is a time that mocks out some methods of time.Time.
+// It doesn't advance the time over time, but only changes it with calls to Set.
+// Use NewMockTime to create Mocktimes, don't instanciate the struct directly unless you want to mess with the sync Cond.
+type MockTime struct {
```

Comment:
Unless there is a good case against it, can you look at using https://github.com/benbjohnson/clock in place of our own. We already have an indirect dependency on his framework.

Reply:
Sounds like a good idea.  When I built the time wrapper, I didn't know about ben's library.  I'll see if I can switch it out.